lamindb/curators/__init__.py

```diff
@@ -519,6 +524,178 @@ def save_artifact(
+
+    Example::
+
+        import lamindb as ln
+        import bionty as bt
+
+        # define obs schema
+        obs_schema = ln.Schema(
+            name="mudata_papalexi21_subset_obs_schema",
+            features=[
+                ln.Feature(name="perturbation", dtype="cat[ULabel[Perturbation]]").save(),
+                ln.Feature(name="replicate", dtype="cat[ULabel[Replicate]]").save(),
+            ],
+        ).save()
+
+        # define the ['rna'].obs schema
+        obs_schema_rna = ln.Schema(
+            name="mudata_papalexi21_subset_rna_obs_schema",
+            features=[
+                ln.Feature(name="nCount_RNA", dtype=int).save(),
+                ln.Feature(name="nFeature_RNA", dtype=int).save(),
+                ln.Feature(name="percent.mito", dtype=float).save(),
+            ],
+            coerce_dtype=True,
+        ).save()
+
+        # define the ['hto'].obs schema
+        obs_schema_hto = ln.Schema(
+            name="mudata_papalexi21_subset_hto_obs_schema",
+            features=[
+                ln.Feature(name="nCount_HTO", dtype=int).save(),
+                ln.Feature(name="nFeature_HTO", dtype=int).save(),
+                ln.Feature(name="technique", dtype=bt.ExperimentalFactor).save(),
+            ],
+            coerce_dtype=True,
+        ).save()
+
+        # define ['rna'].var schema
+        var_schema_rna = ln.Schema(
+            name="mudata_papalexi21_subset_rna_var_schema",
+            itype=bt.Gene.symbol,
+            dtype=float,
+        ).save()
+
+        # define composite schema
+        mudata_schema = ln.Schema(
+            name="mudata_papalexi21_subset_mudata_schema",
+            otype="MuData",
+            components={
+                "obs": obs_schema,
+                "rna:obs": obs_schema_rna,
+                "hto:obs": obs_schema_hto,
+                "rna:var": var_schema_rna,
+            },
+        ).save()
+
+        # curate an AnnData
+        mdata = ln.core.datasets.mudata_papalexi21_subset()
+        bt.settings.organism = "human" # set the organism
+        curator = ln.curators.MuDataCurator(mdata, mudata_schema)
+        artifact = curator.save_artifact(key="example_datasets/mudata_papalexi21_subset.h5mu")
+        assert artifact.schema == mudata_schema
+    """
+
+    def __init__(
+        self,
+        dataset: MuData | Artifact,
+        schema: Schema,
+    ) -> None:
+        super().__init__(dataset=dataset, schema=schema)
+        if not data_is_mudata(self._dataset):
+            raise InvalidArgument("dataset must be MuData-like.")
+        if schema.otype != "MuData":
+            raise InvalidArgument("Schema otype must be 'MuData'.")
+
+        # in form of {slot: DataFrameCurator}
+        self._slots: dict[str, DataFrameCurator] = {}
+        # in form of {modality: var_field}
+        self._var_fields: dict[str, FieldAttr] = {}
+        # in form of {modality: obs_fields}
+        self._obs_fields: dict[str, dict[str, FieldAttr]] = {}
```

Comment:
Could this be inherited or are we really validating modality-specific `.obs`? This might be a design flaw in `MuData` then.

Reply:
Ideally we have only a single `.obs` for the whole `MuData`?

Reply:
We are validating modality-specific obs, I agree we could also get it from `.obs`. But I'm not entirely sure if all modality-specific obs will also show up in global `.obs`. @Zethson Do you know?

Reply:
AFAIK the modality specific obs do NOT show up in the global obs. The global obs are shared among all modalities whereas the modality specific ones are not.

Reply:
But in the ln.core.datasets.mudata_papalexi21_subset(), they in fact do with `modality:` as prefix. I'm a bit confused, as I've also worked with other datasets that don't do that.
<img width="680" alt="Screenshot 2025-03-10 at 17 06 39" src="https://github.com/user-attachments/assets/fe6bfd36-de4c-47ca-9d7d-bef9b48719f5" />

Reply:
Yes, I updated my comment.

Reply:
And in comparison, if I access a modality specific obs, the global obs are not there:
<img width="538" alt="Screenshot 2025-03-10 at 17 07 59" src="https://github.com/user-attachments/assets/edc09a91-ab50-454f-8bd4-0c3c3e3afdd4" />

Reply:
I think having a single `.obs` for mudata is a bit problematic:
1. modality-specific obs columns might not present
2. even when they are present, the column name has `modality:` as a prefix, which we need to manually strip to match feature names. There will also be duplication issues with `rna:x` and `hto:x` both point to the same feature but present in two columns.

So overall I think it's cleaner to model the mudata schema by modalities.

Another option of the composite schema could be:
```
mudata_schema = ln.Schema(
    name="mudata_papalexi21_subset_mudata_schema",
    otype="MuData",
    components={
        "rna": rna_anndata_schema,
        "hto": hto_anndata_schema,
    },
).save()
```
But the issue is then it's difficult to pass the global `.obs` as a component.